In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as pld
import seaborn as sns
import sqlite3

# Exercise 1

Using the same Data Base as in lecture 2, please execute the following:

**1.1** Select unique values in the year column and order them in acsending order. How many unique values are there?

**1.2** Create a column with vehicle age values

**1.3** If a vehicle is older than 45 years, then label it as "45+". Otherwise, label it as "<45". Create a new column that takes these two values

**1.4** How many vehicles are in our dataset that are under 45 and 45 and over?

**1.5** List all unique vehicle name values that have 8 cylinders with horsepower that is greater than 125 but doesn't exceed 200


In [236]:
# 1 
conn = sqlite3.connect('../data/mtcars.sqlite')
df = pd.read_sql_query("SELECT * FROM results", conn)
df

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0    18.0          8         307.0      130.0  3504.0          12.0    70   
1    15.0          8         350.0      165.0  3693.0          11.5    70   
2    18.0          8         318.0      150.0  3436.0          11.0    70   
3    16.0          8         304.0      150.0  3433.0          12.0    70   
4    17.0          8         302.0      140.0  3449.0          10.5    70   
..    ...        ...           ...        ...     ...           ...   ...   
392  27.0          4         140.0       86.0  2790.0          15.6    82   
393  44.0          4          97.0       52.0  2130.0          24.6    82   
394  32.0          4         135.0       84.0  2295.0          11.6    82   
395  28.0          4         120.0       79.0  2625.0          18.6    82   
396  31.0          4         119.0       82.0  2720.0          19.4    82   

     origin                       name  
0         1  chevrolet chevelle malibu  
1         1          buick skylark 320  
2         1         plymouth satellite  
3         1              amc rebel sst  
4         1                ford torino  
..      ...                        ...  
392       1            ford mustang gl  
393       2                  vw pickup  
394       1              dodge rampage  
395       1                ford ranger  
396       1                 chevy s-10  

[397 rows x 9 columns]

In [237]:
# 1.1 Select unique values in the year column and order them in acsending order. How many unique values are there?
pd.read_sql_query("SELECT DISTINCT(year), COUNT(year) as count FROM results GROUP BY year ORDER BY year asc", conn)

year  count
0     70     29
1     71     28
2     72     28
3     73     40
4     74     27
5     75     30
6     76     34
7     77     28
8     78     36
9     79     29
10    80     29
11    81     29
12    82     30

In [252]:
pd.read_sql_query("SELECT COUNT(DISTINCT(year)) AS count FROM results", conn)

count
0     13

There are 13 unique values.

In [238]:
# 1.2 Create a column with vehicle age values
pd.read_sql_query("SELECT *, 123-year as age FROM results",conn)

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0    18.0          8         307.0      130.0  3504.0          12.0    70   
1    15.0          8         350.0      165.0  3693.0          11.5    70   
2    18.0          8         318.0      150.0  3436.0          11.0    70   
3    16.0          8         304.0      150.0  3433.0          12.0    70   
4    17.0          8         302.0      140.0  3449.0          10.5    70   
..    ...        ...           ...        ...     ...           ...   ...   
392  27.0          4         140.0       86.0  2790.0          15.6    82   
393  44.0          4          97.0       52.0  2130.0          24.6    82   
394  32.0          4         135.0       84.0  2295.0          11.6    82   
395  28.0          4         120.0       79.0  2625.0          18.6    82   
396  31.0          4         119.0       82.0  2720.0          19.4    82   

     origin                       name  age  
0         1  chevrolet chevelle malibu   53  
1         1          buick skylark 320   53  
2         1         plymouth satellite   53  
3         1              amc rebel sst   53  
4         1                ford torino   53  
..      ...                        ...  ...  
392       1            ford mustang gl   41  
393       2                  vw pickup   41  
394       1              dodge rampage   41  
395       1                ford ranger   41  
396       1                 chevy s-10   41  

[397 rows x 10 columns]

In [239]:
# 1.3 If a vehicle is older than 45 years, then label it as "45+". 
# Otherwise, label it as "<45". Create a new column that takes these two values
pd.read_sql_query("SELECT *, CASE WHEN 123-year <45 THEN '<45' ELSE '45+' END AS age_grp FROM results",conn)

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0    18.0          8         307.0      130.0  3504.0          12.0    70   
1    15.0          8         350.0      165.0  3693.0          11.5    70   
2    18.0          8         318.0      150.0  3436.0          11.0    70   
3    16.0          8         304.0      150.0  3433.0          12.0    70   
4    17.0          8         302.0      140.0  3449.0          10.5    70   
..    ...        ...           ...        ...     ...           ...   ...   
392  27.0          4         140.0       86.0  2790.0          15.6    82   
393  44.0          4          97.0       52.0  2130.0          24.6    82   
394  32.0          4         135.0       84.0  2295.0          11.6    82   
395  28.0          4         120.0       79.0  2625.0          18.6    82   
396  31.0          4         119.0       82.0  2720.0          19.4    82   

     origin                       name age_grp  
0         1  chevrolet chevelle malibu     45+  
1         1          buick skylark 320     45+  
2         1         plymouth satellite     45+  
3         1              amc rebel sst     45+  
4         1                ford torino     45+  
..      ...                        ...     ...  
392       1            ford mustang gl     <45  
393       2                  vw pickup     <45  
394       1              dodge rampage     <45  
395       1                ford ranger     <45  
396       1                 chevy s-10     <45  

[397 rows x 10 columns]

In [241]:
# 1.4 How many vehicles are in our dataset that are under 45 and 45 and over?
pd.read_sql_query("""SELECT COUNT(*) AS count,
                    CASE WHEN 123-year <45 THEN '<45' ELSE '45+' END AS age_grp FROM results GROUP BY age_grp"""
                  ,conn)

count age_grp
0    280     45+
1    117     <45

In [242]:
# 1.5 List all unique vehicle name values that have 8 cylinders with horsepower that is greater than 125 but doesn't exceed 200
names = pd.read_sql_query("""SELECT DISTINCT(name) FROM results WHERE cylinders=8 AND horsepower>125 AND horsepower<=200"""
                          , conn)
names

name
0              chevrolet chevelle malibu
1                      buick skylark 320
2                     plymouth satellite
3                          amc rebel sst
4                            ford torino
..                                   ...
65                       ford ltd landau
66                 mercury grand marquis
67                       dodge st. regis
68               buick estate wagon (sw)
69  chrysler lebaron town @ country (sw)

[70 rows x 1 columns]

In [96]:
list(names['name'].values)

['chevrolet chevelle malibu',
 'buick skylark 320',
 'plymouth satellite',
 'amc rebel sst',
 'ford torino',
 'ford galaxie 500',
 'amc ambassador dpl',
 'dodge challenger se',
 "plymouth 'cuda 340",
 'chevrolet monte carlo',
 'chevy c20',
 'hi 1200d',
 'chevrolet impala',
 'pontiac catalina brougham',
 'plymouth fury iii',
 'dodge monaco (sw)',
 'ford country squire (sw)',
 'pontiac safari (sw)',
 'pontiac catalina',
 'amc ambassador sst',
 'buick lesabre custom',
 'oldsmobile delta 88 royale',
 'chrysler newport royal',
 'amc matador (sw)',
 'chevrolet chevelle concours (sw)',
 'ford gran torino (sw)',
 'plymouth satellite custom (sw)',
 'buick century 350',
 'amc matador',
 'chevrolet malibu',
 'ford gran torino',
 'dodge coronet custom',
 'mercury marquis brougham',
 'chevrolet caprice classic',
 'ford ltd',
 'plymouth fury gran sedan',
 'amc ambassador brougham',
 'ford country',
 'plymouth custom suburb',
 'oldsmobile vista cruiser',
 'chevrolet monte carlo s',
 'dodge dart custo

# Exercise 2 Songs Database

**2.1** Take the songs dataset and turn it into a single `sqlite` database. 

In [98]:
# 2.1
song = pd.read_csv('../data/song_data.csv')
conn = sqlite3.connect('song.sqlite')
song.to_sql('song_data',con=conn,index=False,if_exists='replace')

18835

In [99]:
pd.read_sql("SELECT name FROM sqlite_master WHERE type='table'",conn)

name
0  song_data

In [100]:
pd.read_sql_query("""SELECT * FROM song_data""",conn)

song_name  song_popularity  song_duration_ms  \
0      Boulevard of Broken Dreams               73            262333   
1                      In The End               66            216933   
2               Seven Nation Army               76            231733   
3                      By The Way               74            216933   
4               How You Remind Me               56            223826   
...                           ...              ...               ...   
18830              Let It Breathe               60            159645   
18831                     Answers               60            205666   
18832      Sudden Love (Acoustic)               23            182211   
18833           Gentle on My Mind               55            352280   
18834                    Up to Me               60            193533   

       acousticness  danceability  energy  instrumentalness  key  liveness  \
0          0.005520         0.496   0.682          0.000029    8    0.0589   
1          0.010300         0.542   0.853          0.000000    3    0.1080   
2          0.008170         0.737   0.463          0.447000    0    0.2550   
3          0.026400         0.451   0.970          0.003550    0    0.1020   
4          0.000954         0.447   0.766          0.000000   10    0.1130   
...             ...           ...     ...               ...  ...       ...   
18830      0.893000         0.500   0.151          0.000065   11    0.1110   
18831      0.765000         0.495   0.161          0.000001   11    0.1050   
18832      0.847000         0.719   0.325          0.000000    0    0.1250   
18833      0.945000         0.488   0.326          0.015700    3    0.1190   
18834      0.911000         0.640   0.381          0.000254    4    0.1040   

       loudness  audio_mode  speechiness    tempo  time_signature  \
0        -4.095           1       0.0294  167.060               4   
1        -6.407           0       0.0498  105.256               4   
2        -7.828           1       0.0792  123.881               4   
3        -4.938           1       0.1070  122.444               4   
4        -5.065           1       0.0313  172.011               4   
...         ...         ...          ...      ...             ...   
18830   -16.107           1       0.0348  113.969               4   
18831   -14.078           0       0.0301   94.286               4   
18832   -12.222           1       0.0355  130.534               4   
18833   -12.020           1       0.0328  106.063               4   
18834   -11.790           1       0.0302   91.490               4   

       audio_valence  
0              0.474  
1              0.370  
2              0.324  
3              0.198  
4              0.574  
...              ...  
18830          0.300  
18831          0.265  
18832          0.286  
18833          0.323  
18834          0.581  

[18835 rows x 15 columns]

# 2.2 Queries

**2.2.1** Use a single query to see if your favorite song is in this dataset

**2.2.2** How many unique popularity scores are there?

**2.2.3** How many song names are in this dataset that contain _'radio edit'_ in their name? Make sure your query is not case sensitive and there is only one single character (space) between _'radio'_ and _'edit'_.

**2.2.4** Write an SQL query to report songs with popularity of 99 and above. How many songs are there?

**2.2.5** Create a new column that contains song duration in the following format m:s. Example, the duration of Boulevard of Broken Dreams is 4:22

**2.2.6** Display the minimum, maximum and average popularity scores. The output of your query should contain only one row with 3 columns.

**2.2.7** Display average popularity scores and count for songs that contain "love" in their name and those that don't. Is there any difference in their popularity? 

**2.2.8** Display top 5 songs with the longest names.



In [224]:
# 2.2.1 Use a single query to see if your favorite song is in this dataset
pd.read_sql_query("""SELECT * FROM song_data WHERE song_name LIKE '%Purple Haze%'""", conn)

song_name  song_popularity  song_duration_ms  acousticness  danceability  \
0  Purple Haze               71            170813       0.00876         0.533   

   energy  instrumentalness  key  liveness  loudness  audio_mode  speechiness  \
0   0.905             0.578    2    0.0698     -5.27           1       0.0754   

   tempo  time_signature  audio_valence  
0  108.9               4          0.486

In [120]:
# 2.2.2 How many unique popularity scores are there?
pd.read_sql_query("""SELECT COUNT(DISTINCT song_popularity) FROM song_data""", conn)

COUNT(DISTINCT song_popularity)
0                              101

In [229]:
# 2.2.3 How many song names are in this dataset that contain 'radio edit' in their name? 
# Make sure your query is not case sensitive and there is only one single character (space) between 'radio' and 'edit'.
pd.read_sql_query("""SELECT COUNT(song_name) FROM song_data WHERE LOWER(song_name) LIKE '%radio edit%'""",conn)

COUNT(song_name)
0               132

In [230]:
# 2.2.4 Write an SQL query to report songs with popularity of 99 and above. How many songs are there?
pd.read_sql_query("""SELECT DISTINCT(song_name) FROM song_data WHERE song_popularity>99""",conn)

song_name
0   Happier

There is one song, Happier.

In [233]:
# 2.2.5 Create a new column that contains song duration in the following format m:s. 
# Example, the duration of Boulevard of Broken Dreams is 4:22
pd.read_sql_query("""SELECT song_name, song_duration_ms,
           (song_duration_ms / 60000) || ':' || strftime('%S', (song_duration_ms / 1000), 'unixepoch') AS time_in_m_s
            FROM song_data
            """,conn)

song_name  song_duration_ms time_in_m_s
0      Boulevard of Broken Dreams            262333        4:22
1                      In The End            216933        3:36
2               Seven Nation Army            231733        3:51
3                      By The Way            216933        3:36
4               How You Remind Me            223826        3:43
...                           ...               ...         ...
18830              Let It Breathe            159645        2:39
18831                     Answers            205666        3:25
18832      Sudden Love (Acoustic)            182211        3:02
18833           Gentle on My Mind            352280        5:52
18834                    Up to Me            193533        3:13

[18835 rows x 3 columns]

In [209]:
# 2.2.6 Display the minimum, maximum and average popularity scores. 
# The output of your query should contain only one row with 3 columns.
pd.read_sql_query("""SELECT MIN(song_popularity) AS minimum, MAX(song_popularity) AS maximum,
                    ROUND(AVG(song_popularity),2) AS average FROM song_data""", conn)

minimum  maximum  average
0        0      100    52.99

In [210]:
# 2.2.7 Display average popularity scores and count for songs that contain "love" in their name and those that don't. 
# Is there any difference in their popularity?
pd.read_sql_query("""SELECT ROUND(AVG(song_popularity),2) AS average FROM song_data WHERE song_name LIKE '%love%'""",conn)

average
0    54.39

In [211]:
pd.read_sql_query("""SELECT ROUND(AVG(song_popularity),2) AS average FROM song_data WHERE song_name NOT LIKE '%love%'""",conn)

average
0    52.94

Yes, the songs that contain 'love' in their name score a higher average song popularity of 54.39 than the ones that don't, with a score of 52.94.

In [213]:
# 2.2.8 Display top 5 songs with the longest names.
long_songs=pd.read_sql_query("""SELECT song_name FROM song_data ORDER BY LENGTH(song_name) DESC LIMIT 5""",conn)
long_songs

song_name
0  Viajo Sin Ver (Remix) [feat. De La Ghetto, Alm...
1  Riff Off: Mickey/Like A Virgin/Hit Me With You...
2  Wu Tang Forever (ft. Ghostface Killah, Raekwon...
3  Bellas Finals: Price Tag/Don’t You (Forget Abo...
4  I Don’t Wanna Live Forever (Fifty Shades Darke...

In [222]:
for i in list(long_songs.song_name.values): print(i,"\n")

Viajo Sin Ver (Remix) [feat. De La Ghetto, Almighty, Miky Woodz, El Alfa, Noriel, Ele A El Dominio, Lyan, Juanka El Problematik, Pusho, Jeycyn] 

Riff Off: Mickey/Like A Virgin/Hit Me With Your Best Shot/S&M/Let’s Talk About Sex/I’ll Make Love To You/Feels Like The First Time/No Diggity 

Wu Tang Forever (ft. Ghostface Killah, Raekwon, RZA, Method Man, Inspectah Deck, Cappadonna, Jackpot Scotty Wotty, U-God, Masta Killa, GZA) 

Bellas Finals: Price Tag/Don’t You (Forget About Me)/Give Me Everything/Just The Way You Are/Party In The U.S.A./Turn The Beat Around 

I Don’t Wanna Live Forever (Fifty Shades Darker) - From "Fifty Shades Darker (Original Motion Picture Soundtrack)" 

